### PR TITLE
[pyzx] add pyzx arc conversion

### PR DESCRIPTION
**Pull request by cqc-melf**

### CODE REVIEW

File: modules/pytket-pyzx/tests/pyzx_convert_test.py [hunk 1]
```diff
@@ -60,3 +69,66 @@ def test_invalid_gate() -> None:
...
+
+    aas_pass = CXMappingPass(arc, g_place)
+
+    aas_pass.apply(c)
+
+    _, pyzx_circ, inv_map = pytket_to_pyzx_placed_circ(c, arc)
+
+    pytket_circ_2 = pyzx_to_pytket_placed_circ(pyzx_circ, inv_map)
+
+    assert pytket_circ_2.qubits == c.qubits
```
Comment (cqc-alec) [inline modules/pytket-pyzx/tests/pyzx_convert_test.py:134]: Can we check for circuit equality?

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 2]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
+    vertices = arcgraph.add_vertices(len(pytket_arc.nodes))
+    arc_dict = dict()
+    qubit_dict = dict()
+
+    for i, x in enumerate(pytket_arc.nodes):
+        q = Qubit(i)
+        qubit_dict[q] = i
+
+    for i, x in enumerate(pytket_arc.nodes):
+        arc_dict[x] = qubit_dict[q_map[x]]
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: That is only possible if the qubits are sorted, or do you mean something else?

File: modules/pytket-pyzx/pytket/extensions/pyzx/__init__.py [hunk 3]
```diff
@@ -15,4 +15,11 @@
...
 # _metadata.py is copied to the folder after installation.
 from ._metadata import __extension_version__, __extension_name__  # type: ignore
-from .pyzx_convert import tk_to_pyzx, pyzx_to_tk
+from .pyzx_convert import (
+    tk_to_pyzx,
+    pyzx_to_tk,
+    pytket_to_pyzx_arc,
+    pyzx_to_pytket_arc,
+    pytket_to_pyzx_placed_circ,
+    pyzx_to_pytket_placed_circ,
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/__init__.py]: done in @cqc-melf
[rename pytket to tk in functionnames](https://github.com/CQCL/pytket-extensions/pull/393/commits/bcb333904b31ca9cb13a782ced7afc33d77f6eb1)
[bcb3339](https://github.com/CQCL/pytket-extensions/pull/393/commits/bcb333904b31ca9cb13a782ced7afc33d77f6eb1)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 4]
```diff
@@ -15,11 +15,14 @@
 """Methods to allow conversion between pyzx and tket data types
 """
 
-from typing import Dict
+from typing import Dict, Tuple
 from fractions import Fraction
 import pyzx as zx  # type: ignore
 from pyzx.circuit import Circuit as pyzxCircuit  # type: ignore
-from pytket.circuit import OpType, Circuit, Op  # type: ignore
+from pyzx.routing.architecture import Architecture as pyzxarc  # type: ignore
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[rename pyzxarc to PyzxArc](https://github.com/CQCL/pytket-extensions/pull/393/commits/edb7cd3aaa426f3a8b55487d2e7eb9a169164e9d)
[edb7cd3](https://github.com/CQCL/pytket-extensions/pull/393/commits/edb7cd3aaa426f3a8b55487d2e7eb9a169164e9d)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 5]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
     return c
+
+
+def pytket_to_pyzx_arc(pytket_arc: Architecture) -> pyzxarc:
+    """
+    Convert a pytket :py:class:`Architecture` to a pyzx
+    :py:class:`pyzx.routing.architecture.Architecture` .
+    The conversion will remove all the node names and will
+    keep them only integer named in the order they are given
+    in the node set of the pytket.
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[rename pyzxgraph to PyzxGraph](https://github.com/CQCL/pytket-extensions/pull/393/commits/84c833c261aa04b9d44cbcefca8ba809416c9947)
[84c833c](https://github.com/CQCL/pytket-extensions/pull/393/commits/84c833c261aa04b9d44cbcefca8ba809416c9947)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 6]
```diff
@@ -15,11 +15,14 @@
...
 """
 
-from typing import Dict
+from typing import Dict, Tuple
 from fractions import Fraction
 import pyzx as zx  # type: ignore
 from pyzx.circuit import Circuit as pyzxCircuit  # type: ignore
-from pytket.circuit import OpType, Circuit, Op  # type: ignore
+from pyzx.routing.architecture import Architecture as pyzxarc  # type: ignore
+from pyzx.graph.graph import Graph as pyzxgraph  # type: ignore
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[rename pyzxgraph to PyzxGraph](https://github.com/CQCL/pytket-extensions/pull/393/commits/84c833c261aa04b9d44cbcefca8ba809416c9947)
[84c833c](https://github.com/CQCL/pytket-extensions/pull/393/commits/84c833c261aa04b9d44cbcefca8ba809416c9947)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[replace in the node set of .](https://github.com/CQCL/pytket-extensions/pull/393/commits/73d062ab69df90daf79e4d12752169ccfa427755)
[73d062a](https://github.com/CQCL/pytket-extensions/pull/393/commits/73d062ab69df90daf79e4d12752169ccfa427755)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 7]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
+        arc_dict[x] = i
+
+    edges = [
+        (vertices[arc_dict[v1]], vertices[arc_dict[v2]])
+        for v1, v2 in pytket_arc.coupling
+    ]
+
+    arcgraph.add_edges(edges)
+
+    pyzx_arc = pyzxarc("pyzx_arc", coupling_graph=arcgraph)
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: have added that for both functions

done in @cqc-melf
[add parameter for src name in pyzx](https://github.com/CQCL/pytket-extensions/pull/393/commits/d12d8b65a6608eac4b021e4bf631846b1adcd361)
[d12d8b6](https://github.com/CQCL/pytket-extensions/pull/393/commits/d12d8b65a6608eac4b021e4bf631846b1adcd361)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 8]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
+    to a pytket :py:class:`Architecture` .
+
+    :param pytket_arc: A Architecture to be converted
+
+    :return: The converted pyzx Architecture
+    """
+
+    pytket_arc = Architecture([tuple(s) for s in pyzx_arc.graph.edges()])
+
+    return pytket_arc
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[return directly in pyzx_to_tk_arc](https://github.com/CQCL/pytket-extensions/pull/393/commits/6a353c50d65090e2d2d41b054185a7030065bcba)
[6a353c5](https://github.com/CQCL/pytket-extensions/pull/393/commits/6a353c50d65090e2d2d41b054185a7030065bcba)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 9]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
+        arc_dict[x] = qubit_dict[q_map[x]]
+
+    edges = [
+        (vertices[arc_dict[v1]], vertices[arc_dict[v2]])
+        for v1, v2 in pytket_arc.coupling
+    ]
+
+    arcgraph.add_edges(edges)
+
+    pyzx_arc = pyzxarc("pyzx_arc", coupling_graph=arcgraph)
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[add parameter for src name in pyzx](https://github.com/CQCL/pytket-extensions/pull/393/commits/d12d8b65a6608eac4b021e4bf631846b1adcd361)
[d12d8b6](https://github.com/CQCL/pytket-extensions/pull/393/commits/d12d8b65a6608eac4b021e4bf631846b1adcd361)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 10]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
+
+
+def pyzx_to_pytket_placed_circ(
+    pyzx_circ: pyzxCircuit, q_map: Dict[UnitID, UnitID]
+) -> Circuit:
+    """
+    Convert a :py:class:`pyzx.Circuit` and a placment map
+    to a placed tket :py:class:`Circuit` .
+    All PyZX basic gate operations are currently supported by pytket. Run
+    `pyzx_circuit_name.to_basic_gates()` before conversion.
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: Removed that for the old function documentation as well, see @cqc-melf
[remove >>all PyZX basic gate operations are currently supported by py…](https://github.com/CQCL/pytket-extensions/pull/393/commits/500807d7467a485567ff783e32484ed934a998ba)
[500807d](https://github.com/CQCL/pytket-extensions/pull/393/commits/500807d7467a485567ff783e32484ed934a998ba)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 11]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
+    :param pytket_circ: A circuit to be converted
+    :param pytket_arc: Corresponding Architecture
+    :param denominator_limit: The limit for denominator size when converting
+        floats to fractions. Smaller limits allow for correct representation of simple
+        fractions with non-exact floating-point representations, while larger limits
+        allow for more precise angles.
+
+    :return: Tuple containing generated :py:class:`pyzx.Circuit` ,
+        :py:class:`pyzx.routing.architecture.Architecture` and
+        map for conversion
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[add more information about the map in the documentation](https://github.com/CQCL/pytket-extensions/pull/393/commits/d2a1a0a2bd894aa6858f5df441c6f29fb5de5330)
[d2a1a0a](https://github.com/CQCL/pytket-extensions/pull/393/commits/d2a1a0a2bd894aa6858f5df441c6f29fb5de5330)

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py [hunk 12]
```diff
@@ -132,3 +135,128 @@ def pyzx_to_tk(pyzx_circ: pyzxCircuit) -> Circuit:
...
+    pyzx_circ: pyzxCircuit, q_map: Dict[UnitID, UnitID]
+) -> Circuit:
+    """
+    Convert a :py:class:`pyzx.Circuit` and a placment map
+    to a placed tket :py:class:`Circuit` .
+    All PyZX basic gate operations are currently supported by pytket. Run
+    `pyzx_circuit_name.to_basic_gates()` before conversion.
+
+    :param pyzx_circ: A circuit to be converted
+    :param q_map: placment map
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: done in @cqc-melf
[add more information to the doc about the map parameter](https://github.com/CQCL/pytket-extensions/pull/393/commits/7b3dc2de28f40b48cf9b6a757552fc2f255a7513)
[7b3dc2d](https://github.com/CQCL/pytket-extensions/pull/393/commits/7b3dc2de28f40b48cf9b6a757552fc2f255a7513)

File: modules/pytket-pyzx/tests/pyzx_convert_test.py [hunk 13]
```diff
@@ -60,3 +69,66 @@ def test_invalid_gate() -> None:
...
+
+    aas_pass = AASRouting(arc)
+
+    aas_pass.apply(c)
+
+    _, pyzx_circ, inv_map = pytket_to_pyzx_placed_circ(c, arc)
+
+    pytket_circ_2 = pyzx_to_pytket_placed_circ(pyzx_circ, inv_map)
+
+    assert pytket_circ_2.qubits == c.qubits
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply modules/pytket-pyzx/tests/pyzx_convert_test.py:105]: I have added a state comparison, is that what you had in mind? done in @cqc-melf
[add check for circuit equality](https://github.com/CQCL/pytket-extensions/pull/393/commits/f34943fb98fed2898b7e6a5d1e07074e75a50277)
[f34943f](https://github.com/CQCL/pytket-extensions/pull/393/commits/f34943fb98fed2898b7e6a5d1e07074e75a50277)

File: modules/pytket-pyzx/tests/pyzx_convert_test.py (hunk 1 above)
Comment (cqc-melf) [reply modules/pytket-pyzx/tests/pyzx_convert_test.py:138]: see above

File: modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cqc-alec) [reply modules/pytket-pyzx/pytket/extensions/pyzx/pyzx_convert.py]: No, ignore me.

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**